File: 13.SlidingWindow滑动窗口/leetcode_3090_maximumLengthSubstring.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int maximumLengthSubstring(string s) {
        int n = s.size();
        int r = 0;
        unordered_map<int, int> cnt;
        int res = 0;
        for (int l=0; l<n; l++){
            while (r<n && cnt[s[r]] <= 1){
                cnt[s[r]]++;
                r++;
            }
            res = max(res, r-l);
            cnt[s[l]]--;
        }
        return res;
    }
};

int main(){
    Solution sl;
    string s = "bcbbbacb";
    cout << sl.maximumLengthSubstring(s) << endl;
}

File: 13.SlidingWindow滑动窗口/leetcode_3090_maximumLengthSubstring.py
from collections import defaultdict

class Solution:
    def maximumLengthSubstring(self, s: str) -> int:
        cnt = defaultdict(int)
        r = 0
        res = 0
        for l in range(len(s)):
            while r<len(s) and cnt[s[r]] <= 1:
                cnt[s[r]] += 1
                r += 1
            res = max(res, r-l)
            cnt[s[l]] -= 1
        return res

if __name__ == "__main__":
    sl = Solution()
    s = "bcbbbcab"
    print(sl.maximumLengthSubstring(s))
